In [1]:
import pandas as pd
import json
import re
from nltk.stem import WordNetLemmatizer
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report


#Carga de datos
train_data = pd.read_csv("../../Data/BasedOnEkman/train_ekman.csv")
test_data = pd.read_csv("../../Data/BasedOnEkman/test_ekman.csv")
valid_data = pd.read_csv("../../Data/BasedOnEkman/valid_ekman.csv")



# Cargar el mapeo de Ekman desde el archivo JSON
with open("../../Data/GoEmotions/ekman_mapping.json", "r") as f:
    ekman_mapping = json.load(f)
ekman_labels = ['anger', 'disgust', 'fear', 'joy', 'sadness', 'surprise', 'neutral']
ekman_to_id = {label: idx for idx, label in enumerate(ekman_labels)}

# --- PREPROCESAMIENTO PARA LEMATIZACIÓN ---
lemmatizer = WordNetLemmatizer()

def preprocess_text(text):
    words = re.findall(r'\b\w+\b', text.lower())
    lemmatized_words = [lemmatizer.lemmatize(word) for word in words]
    return " ".join(lemmatized_words)

# Datos
X_train_raw = train_data['Text'].values
X_test_raw = test_data['Text'].values
y_train = train_data['Emotion'].values
y_test = test_data['Emotion'].values

# Datos lematizados
X_train_lem = [preprocess_text(text) for text in X_train_raw]
X_test_lem = [preprocess_text(text) for text in X_test_raw]


In [2]:
# --- DIRECTORIO DE SALIDA DE LOS PLOTS ---
output_dir = "../../Plots/Experiment1/Ekman/Experiment/"


import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import confusion_matrix

# Función para graficar matriz de confusión
def plot_confusion_matrix(y_true, y_pred, labels, title):
    cm = confusion_matrix(y_true, y_pred, labels=list(range(len(labels))))

    plt.figure(figsize=(10, 8))
    sns.heatmap(
        cm,
        annot=True,
        fmt="d",
        cmap="Blues", 
        xticklabels=labels,
        yticklabels=labels,
        cbar=True
    )
    plt.title(f'Matriz de Confusión - {title}')
    plt.ylabel('Etiqueta Verdadera')
    plt.xlabel('Etiqueta Predicha')
    plt.xticks(rotation=45)
    plt.yticks(rotation=0)
    plt.tight_layout()
    # Guardar archivo
    filename = title.replace(" ", "_").replace("(", "").replace(")", "").lower() + ".png"
    plt.savefig(output_dir + filename)
    plt.close()
    
    

In [3]:
# --- FUNCION GENERAL PARA COMPARAR ---
def train_and_evaluate(model, X_train, y_train, X_test, y_test, label):
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    print(f"\n--- {label} ---")
    classification_report(y_test, y_pred)
    report_dict = classification_report(y_test, y_pred, output_dict=True, zero_division=0)
    f1_macro = report_dict["macro avg"]["f1-score"]
    recall_macro = report_dict["macro avg"]["recall"]
    print(f"Accuracy: {accuracy_score(y_test, y_pred):.4f}")
    print(f"F1 Score (macro avg): {f1_macro:.4f}")
    print(f"Recall Score (macro avg): {recall_macro:.4f}")

    # Mostrar matriz de confusión
    
    plot_confusion_matrix(y_test, y_pred, ekman_labels, label)

In [4]:
# --- VECTORIZADORES ---
vectorizer_raw = TfidfVectorizer(max_features=10000)
vectorizer_stopwords = TfidfVectorizer(max_features=10000, stop_words='english')
vectorizer_lowercase = TfidfVectorizer(max_features=10000, lowercase=True)
vectorizer_strip = TfidfVectorizer(max_features=10000, strip_accents='unicode')
vectorizer_stopwords_lowercase = TfidfVectorizer(max_features=10000, stop_words='english', lowercase=True)
vectorizer_strip_lowercase = TfidfVectorizer(max_features=10000, strip_accents='unicode', lowercase=True)
vectorizer_complete = TfidfVectorizer(max_features=10000, stop_words='english', lowercase=True, strip_accents='unicode')

#TRAIN Y TEST CON Y SIN LEMATIZACIÖN (TF-IDF sin hiperparámetros)
X_train_raw_tfidf = vectorizer_raw.fit_transform(X_train_raw)
X_test_raw_tfidf = vectorizer_raw.transform(X_test_raw)

X_train_lem_tfidf = vectorizer_raw.fit_transform(X_train_lem)
X_test_lem_tfidf = vectorizer_raw.transform(X_test_lem)

X_train_lem_tfidf_stopwords = vectorizer_stopwords.fit_transform(X_train_lem)
X_test_lem_tfidf_stopwords = vectorizer_stopwords.transform(X_test_lem)

X_train_lem_tfidf_lowercase = vectorizer_lowercase.fit_transform(X_train_lem)
X_test_lem_tfidf_lowercase = vectorizer_lowercase.transform(X_test_lem)

X_train_lem_tfidf_strip = vectorizer_strip.fit_transform(X_train_lem)
X_test_lem_tfidf_strip = vectorizer_strip.transform(X_test_lem)

X_train_lem_tfidf_stopwords_lowercase = vectorizer_stopwords_lowercase.fit_transform(X_train_lem)
X_test_lem_tfidf_stopwords_lowercase = vectorizer_stopwords_lowercase.transform(X_test_lem)

X_train_lem_tfidf_strip_lowercase = vectorizer_strip_lowercase.fit_transform(X_train_lem)
X_test_lem_tfidf_strip_lowercase = vectorizer_strip_lowercase.transform(X_test_lem)

X_train_lem_tfidf_complete = vectorizer_complete.fit_transform(X_train_lem)
X_test_lem_tfidf_complete = vectorizer_complete.transform(X_test_lem)





In [5]:
#Modelos de clasificación
models = {
    'Logistic Regression': LogisticRegression(max_iter=1000, random_state=42),
    'SVM (linear)': SVC(kernel='linear', class_weight='balanced'),
    'Decision Tree': DecisionTreeClassifier(random_state=42)
}

SoporteVectorial = SVC(kernel='linear', class_weight='balanced')


In [ ]:
# --- COMPARATIVA FINAL ---
print("\n-- TF-IDF --")
for model_name, model in models.items():
    train_and_evaluate(model, X_train_raw_tfidf, y_train, X_test_raw_tfidf, y_test, f"{model_name} (TF-IDF)")


print("\n-- CON LEMATIZAR --")
for model_name, model in models.items():
    train_and_evaluate(model, X_train_lem_tfidf, y_train, X_test_lem_tfidf, y_test, f"{model_name} (TF-IDF Lematizado)")

print("\nEl experimento con el TF-IDF con lematizar obtuvo mejores resultados\n")
print("Es por ello por lo que se comienza la experimientación con el TF-IDF hiperparametrizado para comprobar cuál es la mejor configuración del mismo\n")
print("Además, el clasificador con mejor desempeño fue el SVM linear, por lo que será el usado en los próximos experimentos")


print("\n--TF-IDF SIN STOPWORDS --")
train_and_evaluate(SoporteVectorial, X_train_lem_tfidf_stopwords, y_train, X_test_lem_tfidf_stopwords, y_test, "SVM (linear) (TF-IDF sin stopwords)")

print("\n-- TF-IDF EN MINUSCULAS --")
train_and_evaluate(SoporteVectorial, X_train_lem_tfidf_lowercase, y_train, X_test_lem_tfidf_lowercase, y_test, "SVM (linear) (TF-IDF en minusculas)")

print("\n-- TF-IDF SIN ACENTOS --")
train_and_evaluate(SoporteVectorial, X_train_lem_tfidf_strip, y_train, X_test_lem_tfidf_strip, y_test, "SVM (linear) (TF-IDF sin acentos)")

print("\n-- TF-IDF SIN STOPWORDS Y EN MINUSCULAS --")
train_and_evaluate(SoporteVectorial, X_train_lem_tfidf_stopwords_lowercase, y_train, X_test_lem_tfidf_stopwords_lowercase, y_test, "SVM (linear)  (TF-IDF sin stopwords y en minusculas)")

print("\n-- TF-IDF SIN ACENTOS Y EN MINUSCULAS --")
train_and_evaluate(SoporteVectorial, X_train_lem_tfidf_strip_lowercase, y_train, X_test_lem_tfidf_strip_lowercase, y_test, "SVM (linear)  (TF-IDF sin acentos y en minusculas)")

print("\n-- TF-IDF COMPLETO --")
train_and_evaluate(SoporteVectorial, X_train_lem_tfidf_complete, y_train, X_test_lem_tfidf_complete, y_test, "SVM (linear)  (TF-IDF completo)")






-- TF-IDF --

--- Logistic Regression (TF-IDF) ---
Accuracy: 0.6329
F1 Score (macro avg): 0.4895
Recall Score (macro avg): 0.4302

--- SVM (linear) (TF-IDF) ---
Accuracy: 0.5906
F1 Score (macro avg): 0.5212
Recall Score (macro avg): 0.5703

--- Decision Tree (TF-IDF) ---
Accuracy: 0.5280
F1 Score (macro avg): 0.3915
Recall Score (macro avg): 0.3757

-- CON LEMATIZAR --

--- Logistic Regression (TF-IDF Lematizado) ---
Accuracy: 0.6379
F1 Score (macro avg): 0.5052
Recall Score (macro avg): 0.4412

--- SVM (linear) (TF-IDF Lematizado) ---
Accuracy: 0.5902
F1 Score (macro avg): 0.5232
Recall Score (macro avg): 0.5758

--- Decision Tree (TF-IDF Lematizado) ---
Accuracy: 0.5209
F1 Score (macro avg): 0.3814
Recall Score (macro avg): 0.3630

El experimento con el TF-IDF con lematizar obtuvo mejores resultados

Es por ello por lo que se comienza la experimientación con el TF-IDF hiperparametrizado para comprobar cuál es la mejor configuración del mismo

Además, el clasificador con mejor desemp


-- TF-IDF --

--- Logistic Regression (TF-IDF) ---
Accuracy: 0.6329
F1 Score (macro avg): 0.4895
Recall Score (macro avg): 0.4302

--- SVM (linear) (TF-IDF) ---
Accuracy: 0.5906
F1 Score (macro avg): 0.5212
Recall Score (macro avg): 0.5703

--- Decision Tree (TF-IDF) ---
Accuracy: 0.5280
F1 Score (macro avg): 0.3915
Recall Score (macro avg): 0.3757

-- CON LEMATIZAR --

--- Logistic Regression (TF-IDF Lematizado) ---
Accuracy: 0.6379
F1 Score (macro avg): 0.5052
Recall Score (macro avg): 0.4412

--- SVM (linear) (TF-IDF Lematizado) ---
Accuracy: 0.5902
F1 Score (macro avg): 0.5232
Recall Score (macro avg): 0.5758

--- Decision Tree (TF-IDF Lematizado) ---
Accuracy: 0.5209
F1 Score (macro avg): 0.3814
Recall Score (macro avg): 0.3630

El experimento con el TF-IDF con lematizar obtuvo mejores resultados

Es por ello por lo que se comienza la experimientación con el TF-IDF hiperparametrizado para comprobar cuál es la mejor configuración del mismo

Además, el clasificador con mejor desempeño fue el SVM linear, por lo que será el usado en los próximos experimentos

--TF-IDF SIN STOPWORDS --

--- SVM (linear) (TF-IDF sin stopwords) ---
Accuracy: 0.5801
F1 Score (macro avg): 0.5058
Recall Score (macro avg): 0.5491

-- TF-IDF EN MINUSCULAS --

--- SVM (linear) (TF-IDF en minusculas) ---
Accuracy: 0.5902
F1 Score (macro avg): 0.5232
Recall Score (macro avg): 0.5758

-- TF-IDF SIN ACENTOS --

--- SVM (linear) (TF-IDF sin acentos) ---
Accuracy: 0.5914
F1 Score (macro avg): 0.5229
Recall Score (macro avg): 0.5738

-- TF-IDF SIN STOPWORDS Y EN MINUSCULAS --

--- SVM (linear)  (TF-IDF sin stopwords y en minusculas) ---
Accuracy: 0.5801
F1 Score (macro avg): 0.5058
Recall Score (macro avg): 0.5491

-- TF-IDF SIN ACENTOS Y EN MINUSCULAS --

--- SVM (linear)  (TF-IDF sin acentos y en minusculas) ---
Accuracy: 0.5914
F1 Score (macro avg): 0.5229
Recall Score (macro avg): 0.5738

-- TF-IDF COMPLETO --

--- SVM (linear)  (TF-IDF completo) ---
Accuracy: 0.5797
F1 Score (macro avg): 0.5069
Recall Score (macro avg): 0.5520

-- TF-IDF --

--- Logistic Regression (TF-IDF) ---
Accuracy: 0.6329
              precision    recall  f1-score   support

           0       0.57      0.30      0.39       572
           1       0.69      0.26      0.38        76
           2       0.88      0.29      0.43        80
           3       0.76      0.77      0.77      1863
           4       0.70      0.34      0.46       283
           5       0.56      0.26      0.36       488
           6       0.54      0.79      0.64      1606

    accuracy                           0.63      4968
   macro avg       0.67      0.43      0.49      4968
weighted avg       0.64      0.63      0.61      4968


--- SVM (linear) (TF-IDF) ---
Accuracy: 0.6449
              precision    recall  f1-score   support

           0       0.57      0.32      0.41       572
           1       0.57      0.38      0.46        76
           2       0.76      0.51      0.61        80
           3       0.81      0.75      0.78      1863
           4       0.70      0.46      0.55       283
           5       0.59      0.25      0.35       488
           6       0.54      0.81      0.65      1606

    accuracy                           0.64      4968
   macro avg       0.65      0.50      0.54      4968
weighted avg       0.66      0.64      0.63      4968


--- Decision Tree (TF-IDF) ---
Accuracy: 0.5280
              precision    recall  f1-score   support

           0       0.33      0.29      0.31       572
           1       0.33      0.28      0.30        76
           2       0.33      0.23      0.27        80
           3       0.68      0.69      0.69      1863
           4       0.43      0.31      0.36       283
           5       0.31      0.26      0.28       488
           6       0.50      0.56      0.53      1606

    accuracy                           0.53      4968
   macro avg       0.41      0.38      0.39      4968
weighted avg       0.52      0.53      0.52      4968


-- CON LEMATIZAR --

--- Logistic Regression (TF-IDF Lematizado) ---
Accuracy: 0.6379
              precision    recall  f1-score   support

           0       0.59      0.30      0.40       572
           1       0.80      0.26      0.40        76
           2       0.96      0.33      0.49        80
           3       0.77      0.77      0.77      1863
           4       0.71      0.35      0.47       283
           5       0.55      0.28      0.38       488
           6       0.54      0.79      0.64      1606

    accuracy                           0.64      4968
   macro avg       0.70      0.44      0.51      4968
weighted avg       0.65      0.64      0.62      4968


--- SVM (linear) (TF-IDF Lematizado) ---
Accuracy: 0.6443
              precision    recall  f1-score   support

           0       0.58      0.32      0.41       572
           1       0.56      0.37      0.44        76
           2       0.76      0.53      0.62        80
           3       0.81      0.75      0.78      1863
           4       0.69      0.45      0.54       283
           5       0.57      0.25      0.34       488
           6       0.54      0.82      0.65      1606

    accuracy                           0.64      4968
   macro avg       0.64      0.50      0.54      4968
weighted avg       0.66      0.64      0.63      4968


--- Decision Tree (TF-IDF Lematizado) ---
Accuracy: 0.5209
              precision    recall  f1-score   support

           0       0.30      0.26      0.28       572
           1       0.42      0.30      0.35        76
           2       0.33      0.17      0.23        80
           3       0.68      0.70      0.69      1863
           4       0.38      0.30      0.34       283
           5       0.28      0.24      0.26       488
           6       0.50      0.56      0.53      1606

    accuracy                           0.52      4968
   macro avg       0.41      0.36      0.38      4968
weighted avg       0.51      0.52      0.51      4968


El experimento con el TF-IDF sin lematizar obtuvo mejores resultados

Es por ello por lo que se comienza la experimientación con el TF-IDF hiperparametrizado para comprobar cuál es la mejor configuración del mismo

Además, el clasificador con mejor desempeño fue el SVM linear, por lo que será el usado en los próximos experimentos

--TF-IDF SIN STOPWORDS --

--- SVM (linear) (TF-IDF sin stopwords) ---
Accuracy: 0.6381
              precision    recall  f1-score   support

           0       0.57      0.33      0.42       572
           1       0.57      0.41      0.48        76
           2       0.70      0.53      0.60        80
           3       0.79      0.75      0.77      1863
           4       0.67      0.45      0.54       283
           5       0.66      0.17      0.27       488
           6       0.53      0.81      0.64      1606

    accuracy                           0.64      4968
   macro avg       0.64      0.49      0.53      4968
weighted avg       0.66      0.64      0.62      4968


-- TF-IDF EN MINUSCULAS --

--- SVM (linear) (TF-IDF en minusculas) ---
Accuracy: 0.6449
              precision    recall  f1-score   support

           0       0.57      0.32      0.41       572
           1       0.57      0.38      0.46        76
           2       0.76      0.51      0.61        80
           3       0.81      0.75      0.78      1863
           4       0.70      0.46      0.55       283
           5       0.59      0.25      0.35       488
           6       0.54      0.81      0.65      1606

    accuracy                           0.64      4968
   macro avg       0.65      0.50      0.54      4968
weighted avg       0.66      0.64      0.63      4968


-- TF-IDF SIN ACENTOS --

--- SVM (linear) (TF-IDF sin acentos) ---
Accuracy: 0.6439
              precision    recall  f1-score   support

           0       0.58      0.31      0.41       572
           1       0.57      0.38      0.46        76
           2       0.74      0.53      0.61        80
           3       0.81      0.75      0.78      1863
           4       0.70      0.46      0.55       283
           5       0.58      0.25      0.35       488
           6       0.54      0.81      0.65      1606

    accuracy                           0.64      4968
   macro avg       0.64      0.50      0.54      4968
weighted avg       0.66      0.64      0.63      4968


-- TF-IDF SIN STOPWORDS Y EN MINUSCULAS --

--- SVM (linear)  (TF-IDF sin stopwords y en minusculas) ---
Accuracy: 0.6381
              precision    recall  f1-score   support

           0       0.57      0.33      0.42       572
           1       0.57      0.41      0.48        76
           2       0.70      0.53      0.60        80
           3       0.79      0.75      0.77      1863
           4       0.67      0.45      0.54       283
           5       0.66      0.17      0.27       488
           6       0.53      0.81      0.64      1606

    accuracy                           0.64      4968
   macro avg       0.64      0.49      0.53      4968
weighted avg       0.66      0.64      0.62      4968


-- TF-IDF SIN ACENTOS Y EN MINUSCULAS --

--- SVM (linear)  (TF-IDF sin acentos y en minusculas) ---
Accuracy: 0.6439
              precision    recall  f1-score   support

           0       0.58      0.31      0.41       572
           1       0.57      0.38      0.46        76
           2       0.74      0.53      0.61        80
           3       0.81      0.75      0.78      1863
           4       0.70      0.46      0.55       283
           5       0.58      0.25      0.35       488
           6       0.54      0.81      0.65      1606

    accuracy                           0.64      4968
   macro avg       0.64      0.50      0.54      4968
weighted avg       0.66      0.64      0.63      4968


-- TF-IDF COMPLETO --

--- SVM (linear)  (TF-IDF completo) ---
Accuracy: 0.6361
              precision    recall  f1-score   support

           0       0.58      0.33      0.42       572
           1       0.58      0.38      0.46        76
           2       0.68      0.53      0.59        80
           3       0.79      0.75      0.77      1863
           4       0.67      0.45      0.54       283
           5       0.65      0.17      0.27       488
           6       0.53      0.81      0.64      1606

    accuracy                           0.64      4968
   macro avg       0.64      0.49      0.53      4968
weighted avg       0.65      0.64      0.62      4968